src/components/NewBranchDialog.tsx

```diff
@@ -0,0 +1,440 @@
+
+    // When the HEAD changes, decent probability that we've switched branches:
+    this.props.model.headChanged.connect(this._syncState, this);
+
+    // When the status changes, we may have checked out a new branch (e.g., via the command-line and not via the extension):
+    this.props.model.statusChanged.connect(this._syncState, this);
+  }
+
+  /**
+   * Removes model listeners.
+   */
+  private _removeListeners(): void {
+    this.props.model.repositoryChanged.disconnect(this._syncState, this);
+    this.props.model.headChanged.disconnect(this._syncState, this);
+    this.props.model.statusChanged.disconnect(this._syncState, this);
+  }
+
+  /**
+   * Syncs the component state with the underlying model.
+   */
+  private _syncState(): void {
+    const repo = this.props.model.pathRepository;
+    this.setState({
+      base: repo ? this.state.base : '',
+      current: repo ? this.props.model.currentBranch.name : '',
+      branches: repo ? this.props.model.branches : []
+    });
+  }
+
+  /**
+   * Callback invoked upon closing the dialog.
+   *
+   * @param event - event object
+   */
+  private _onClose = (): void => {
+    this.props.onClose();
+    this.setState({
+      name: '',
+      filter: ''
+    });
+  };
+
+  /**
+   * Callback invoked upon a change to the menu filter.
+   *
+   * @param event - event object
+   */
+  private _onFilterChange = (event: any): void => {
+    this.setState({
+      filter: event.target.value
+    });
+  };
+
+  /**
+   * Callback invoked to reset the menu filter.
+   */
+  private _resetFilter = (): void => {
+    this.setState({
+      filter: ''
+    });
+  };
+
+  /**
+   * Returns a callback which is invoked upon clicking a branch name.
+   *
+   * @param branch - branch name
+   * @returns callback
+   */
+  private _onBranchClickFactory(branch: string) {
+    const self = this;
+    return onClick;
+
+    /**
+     * Callback invoked upon clicking a branch name.
+     *
+     * @private
+     * @param event - event object
+     */
+    function onClick() {
+      self.setState({
+        base: branch
+      });
+    }
+  }
+
+  /**
+   * Callback invoked upon a change to the branch name input element.
+   *
+   * @param event - event object
+   */
+  private _onNameChange = (event: any): void => {
+    this.setState({
+      name: event.target.value
+    });
+  };
+
+  /**
+   * Callback invoked upon clicking a button to create a new branch.
+   *
+   * @param event - event object
+   */
+  private _onCreate = (): void => {
+    const branch = this.state.name;
+
+    // Close the branch dialog:
+    this.props.onClose();
+
+    // Reset the branch name and filter:
+    this.setState({
+      name: '',
+      filter: ''
+    });
+
+    // Create the branch:
+    this._createBranch(branch);
+  };
+
+  /**
+   * Creates a new branch.
+   *
+   * @param branch - branch name
+   */
+  private _createBranch(branch: string): void {
```

Comment:
Can be simplified by using a asynchronous method as for the method in the branch menu.

Reply:
Similar reply as [here](https://github.com/jupyterlab/jupyterlab-git/pull/518#discussion_r370911838).